pep-XXXX.rst

```diff
@@ -0,0 +1,258 @@
+    "libxrender1", "libice6", "libsm6", "libgl1-mesa-glx", and "libglib2.0-0".
+
+On RPM-based systems, these libraries are provided by the packages ::
+
+    "ncurses", "libgcc", "libstdc++", "glibc", "libXext", "libXrender",
+    "libICE", "libSM", "mesa-libGL", and "glib2".
+
+This list was compiled by checking the external shared library dependencies of
+the Canopy [1]_ and Anaconda [2]_ distributions, which both include a wide array
+of the most popular Python modules and have been confirmed in practice to work
+across a wide swath of Linux systems in the wild.
+
+For dependencies on externally-provided versioned symbols in the above shared
+libraries, the following symbol versions are permitted: ::
+
+    GLIBC <= 2.5
+    CXXABI <= 3.4.8
+    GLIBCXX <= 3.4.9
+    GCC <= 4.2.0
+
+These symbol versions were determined by inspecting the latest symbol version
+provided in the libraries distributed with CentOS 5, a Linux distribution
+released in April 2007. In practice, this means that Python wheels which conform
+to this policy should function on almost any linux distribution released after
+this date.
+
+
+Compilation and Tooling
+=======================
+
+To support the compilation of wheels meeting the ``manylinux1`` standard, we
+provide initial drafts of two tools.
+
+The first is a Docker image based on CentOS 5.11, which is recommended as an
+easy to use self-contained build box for compiling  ``manylinux1`` wheels.
+Compiling on a more recently-released linux distribution will generally
+introduce dependencies on too-new versioned symbols. The image comes with a
+full compiler suite installed (``gcc``, ``g++``, and ``gfortran`` 4.8.2) as
+well as the latest releases of Python and pip.
+
+The second tool is a command line executable called ``auditwheel``. First, it
+inspects all of the ELF files inside a wheel to check for dependencies on
+versioned symbols or external shared libraries, and verifies conformance with
+the ``manylinux1`` policy. This includes the ability to add the new platform
+tag to conforming wheels.
+
+In addition, ``auditwheel`` has the ability to automatically modify wheels that
+depend on external shared libraries by copying those shared libraries from
+the system into the wheel itself, and modifying the appropriate RPATH entries
+such that these libraries will be picked up at runtime. This accomplishes a
+similar result as if the libraries had been statically linked without requiring
+changes to the build system.
+
+Neither of these tools are necessary to build wheels which conform with the
+``manylinux1`` policy. Similar results can usually be achieved by statically
+linking external dependencies and/or using certain inline assembly constructs
+to instruct the linker to prefer older symbol versions, however these tricks
+can be quite esoteric.
+
+
+Platform Detection in ``pip``
+=============================
+
+TODO How does ``pip`` detect that it's running on a ``manylinux1`` compatible
+system?
```

Comment:
Suggest text:

```
Because the ``manylinux1`` profile is already known to work for the many thousands of users of popular commercial Python distributions, we suggest that installation tools like ``pip`` should error on the side of assuming that a system *is* compatible, unless there is specific reason to think otherwise.

We know of three main sources of potential incompatibility that are likely to arise in practice:

* A linux distribution that is too old (e.g. RHEL 4)
* A linux distribution that does not use glibc (e.g. Alpine Linux, which is based on musl libc, or Android)
* Eventually, in the future, there may exist distributions that break compatibility with this profile

To handle the first two cases, we propose the following simple and reliable check::

    def have_glibc_version(major, minimum_minor):
        import ctypes

        process_namespace = ctypes.CDLL(None)
        try:
            gnu_get_libc_version = process_namespace.gnu_get_libc_version
        except AttributeError:
            # We are not linked to glibc.
            return False

        gnu_get_libc_version.restype = ctypes.c_char_p
        version_str = gnu_get_libc_version()
        # py2 / py3 compatibility:
        if not isinstance(version_str, str):
            version_str = version_str.decode("ascii")

        version = [int(piece) for piece in version_str.split(".")]
        assert len(version) == 2
        if major != version[0]:
            return False
        if minimum_minor > version[1]:
            return False
        return True

    # CentOS 5 uses glibc 2.5.
    is_manylinux1_compatible = have_glibc_version(2, 5)

To handle the third case, we propose the creation of a file `/etc/python/compatibility.cfg` in ConfigParser format, with sample contents::

   [manylinux1]
   compatible = true

where the supported values for the manylinux1.compatible entry are the same as those supported by the ConfigParser getboolean method.

The proposed logic for ``pip`` or related tools, then, is:

0) If ``distutils.util.get_platform()`` does not start with the string ``"linux"``, then assume the current system is not ``manylinux1`` compatible.
1) If /etc/python/compatibility.conf exists and contains a ``manylinux1`` key, then trust that.
2) Otherwise, if ``have_glibc_version(2, 5)`` returns true, then assume the current system can handle ``manylinux1`` wheels.
3) Otherwise, assume that the current system cannot handle ``manylinux1`` wheels.
```